### PR TITLE
Allow custom type objects to trigger value updates

### PR DESCRIPTION
**Pull request by Artusamak**
Custom types are not correctly supported due to the fact that `UnitOfWork::computeOrRecomputeChangeSet()` uses` === `operations without checking the objects inner values.
This leads to missed update values.

The added test `testValueObjectChangeSets` is red if the changeset detection fragment is not altered.

There is also a `@TODO` added to track the suggestion to use `\Doctrine\Common\Comparable` interface. The problem is that this interface is part of doctrine/commons which is not required by the package at the moment.

|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | #1482

### CODE REVIEW

File: tests/Documents/Functional/DocumentWithCustomTypeValueObject.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+namespace Documents\Functional;
+
+use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
+
+#[ODM\Document]
+class DocumentWithCustomTypeValueObject {
+
+  /** @var string|null */
+  #[ODM\Id]
+  public $id;
```
Comment (GromNaN) [inline tests/Documents/Functional/DocumentWithCustomTypeValueObject.php]: Nitpick: you should use typed properties when adding examples.
```suggestion
  #[ODM\Id]
  public ?string $id;
```

File: tests/Documents/Functional/ValueObjectChild.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+namespace Documents\Functional;
+
+class ValueObjectChild {
+
+  public function __construct(
+    public int $prop1,
+    public int $prop2,
+  ) {}
+
+}
```
Comment (GromNaN) [inline tests/Documents/Functional/ValueObjectChild.php]:
```suggestion
}

```

File: src/UnitOfWork.php [hunk 3]
```diff
@@ -765,16 +765,14 @@ private function computeOrRecomputeChangeSet(ClassMetadata $class, object $docum
...
-                        continue;
-                    }
+                // Target object behaviors.
+                if (is_object($actualValue)) {
+                  if ($actualValue instanceof PersistentCollectionInterface && (! $actualValue->isDirty() && ! $this->isCollectionScheduledForDeletion($actualValue))) {
+                    // consider dirty collections as changed as well
+                    continue;
+                  }
+                  // @TODO: Consider using Comparable interface to delegate
+                  // object comparaisons.
```
Comment (GromNaN) [inline src/UnitOfWork.php]: We don't leave todo in the codebase. Is it something you want to address before the PR is merged?

**Review by GromNaN — COMMENTED** [review]
Could you have a look at the failing tests?

File: src/UnitOfWork.php (hunk 3 above)
Comment (Artusamak) [reply src/UnitOfWork.php]: Yes!

**Comment by Artusamak** [comment]
You've been faster light for review! Thanks. 

Is it considered to add a `.editorfile` matching your code style expectations? Fixing manually is not the easiest solution 😅.

**Comment by Artusamak** [comment]
Thanks to SPL there is now a way to detect the object content update which should do the trick.

File: src/UnitOfWork.php [hunk 4]
```diff
@@ -48,6 +48,7 @@
 use function method_exists;
 use function preg_match;
 use function serialize;
+use function spl_object_hash;
```
Comment (GromNaN) [inline src/UnitOfWork.php:51]: `spl_object_hash($a) === spl_object_hash($b)` is functionally identical to `$a === $b` for two simultaneously live objects ([PHP docs](https://www.php.net/spl_object_hash): the hash can only be reused after the object is destroyed). This import can be dropped.

File: src/UnitOfWork.php [hunk 5]
```diff
@@ -765,16 +766,19 @@ private function computeOrRecomputeChangeSet(ClassMetadata $class, object $docum
...
+                          // consider dirty collections as changed as well
+                          continue;
                     }
 
-                    if (! $actualValue->isDirty() && ! $this->isCollectionScheduledForDeletion($actualValue)) {
-                        // consider dirty collections as changed as well
+                    // If object content has not been modified, nothing to do.
+                    if (is_object($orgValue) && spl_object_hash($actualValue) === spl_object_hash($orgValue)) {
                         continue;
                     }
```
Comment (GromNaN) [inline src/UnitOfWork.php:779]: This restructuring introduces a regression for dirty `ReferenceMany` collections. When a `PersistentCollection` is dirty but its instance is the same object stored in `$originalDocumentData` (`$orgValue === $actualValue` — the normal case), the new code falls through to line 777 and executes `continue`, silently skipping the dirty collection instead of detecting it as changed.

`EmbedMany` collections are unaffected by luck — they go through the association changeset loop (~line 870). `ReferenceMany` collections are explicitly excluded there (`if (isset($mapping['reference'])) { continue; }`), so this block is their only detection path.

PR #3021 adds a test that demonstrates this: it passes on `2.17.x` and fails with this patch applied.

Also, the comment `// consider dirty collections as changed as well` is now on the `continue` of the **clean** collection branch — the opposite of what it says.

Since `spl_object_hash` is equivalent to `===` for live objects, it does not help with value comparison either. I'd suggest keeping the original structure and adding a `==` check as a new `elseif`. PHP's `==` compares objects by value: same class and all properties (public **and** private) equal — https://3v4l.org/sBNOH:

```php
// skip if value has not changed
if ($orgValue === $actualValue) {
    if (! $actualValue instanceof PersistentCollectionInterface) {
        continue;
    }

    if (! $actualValue->isDirty() && ! $this->isCollectionScheduledForDeletion($actualValue)) {
        // consider dirty collections as changed as well
        continue;
    }
} elseif (is_object($orgValue) && is_object($actualValue) && $orgValue == $actualValue) {
    continue;
}
```

This preserves the dirty-collection behaviour and fixes the value object case without any new dependency. The `TODO` on `Comparable` remains valid for edge cases where `==` is not sufficient, but is not a blocker.

File: tests/Tests/Functional/CustomTypeTest.php [hunk 6]
```diff
@@ -216,3 +243,7 @@ public function convertToPHPValue($value): ?Language
 class CustomTypeWithoutClosureToPHP extends Type
 {
 }
+
+class CustomValueObjectChild extends Type
```
Comment (GromNaN) [inline tests/Tests/Functional/CustomTypeTest.php:247]: `CustomValueObjectChild` implements no `convertToPHPValue`/`convertToDatabaseValue`, so the test doesn't quite cover a realistic custom type scenario where the type actually converts values between PHP and MongoDB.

File: tests/Tests/Functional/CustomTypeTest.php [hunk 7]
```diff
@@ -54,6 +57,30 @@ public function testCustomTypeValueConversions(): void
...
+
+        $this->uow->commit();
+        $valueChild  = new ValueObjectChild(12, 12);
+        $root->child = $valueChild;
+
+        $this->uow->computeChangeSets();
+        $changeSet = $this->uow->getDocumentChangeSet($root);
+
+        self::assertArrayHasKey('child', $changeSet);
+        self::assertEquals('12', $changeSet['child'][1]->prop1);
```
Comment (GromNaN) [inline tests/Tests/Functional/CustomTypeTest.php:81]: `prop1` is declared `int`, so this should compare against `12` (no quotes). The assertion should use `assertSame`.

**Review by GromNaN — CHANGES_REQUESTED** [review]
Thanks for the contribution and the test that clearly reproduces the issue! I have a few comments below.
